crates/ecs/src/filter.rs

```diff
@@ -100,141 +90,171 @@ macro_rules! binary_filter_operation {
         ///
         #[doc = concat!(
             "# Example\n```\n",
-            "# use {ecs::filter::With, ecs::filter::", stringify!($name), "};\n",
-            "# use ecs::Read;\n",
+            "# use ecs::filter::{With, ", stringify!($name), "};\n",
+            "# use ecs::systems::Read;\n",
             "# #[derive(Debug)]\n",
             "# struct Position;\n",
+            "# #[derive(Debug)]\n",
             "# struct Player;\n",
+            "# #[derive(Debug)]\n",
             "# struct Enemy;\n",
             "fn player_", $op_name_lowercase, "_enemy(position: Read<Position>, _: ", stringify!($name), "<With<Player>, With<Enemy>>) {\n",
             "    println!(\"An entity at position {:?} is a player ", $op_name_lowercase ," an enemy.\", position);\n",
             "}\n```",
         )]
         #[derive(Debug)]
-        pub struct $name<L: SystemParameter, R: SystemParameter> {
+        pub struct $name<L: Filter, R: Filter> {
             left: PhantomData<L>,
             right: PhantomData<R>,
         }
 
-        impl<L: SystemParameter, R: SystemParameter> SystemParameter for $name<L, R> {
-            type BorrowedData<'components> = (
-                <L as SystemParameter>::BorrowedData<'components>,
-                <R as SystemParameter>::BorrowedData<'components>,
-            );
+        impl<L: Filter, R: Filter> Filter for $name<L, R> {}
+        impl<L: Filter + SystemParameter, R: Filter + SystemParameter> SystemParameter for $name<L, R> {
```

Comment:
The problem with that is it would leak the crate-private `SystemParameter` trait.

Reply:
Ah, I see. But why does `Filter` need to be public when `SystemParameter` is not? Couldn't they both be crate-private, then?

Reply:
Oh, nevermind I just tried it and that's not possible since the filters themselves are public and use it as a type parameter bound. 

Alright, you can resolve this 👍

Reply:
`Filter` needs to be public because the binary/unary filters take `Filter` as a generic parameter.
`SystemParameter` can be private because the `add_system` only require that  `SystemParameters`  is implemeneted, so `SystemParameter` can be private.